src/xpub.cpp

```diff
@@ -106,18 +107,23 @@ void zmq::xpub_t::xread_activated (pipe_t *pipe_)
                 pending_metadata.push_back (metadata);
                 pending_flags.push_back (0);
             } else {
-                bool unique;
-                if (*data == 0)
-                    unique = subscriptions.rm (data + 1, size - 1, pipe_);
-                else
-                    unique = subscriptions.add (data + 1, size - 1, pipe_);
-
-                //  If the (un)subscription is not a duplicate store it so that it can be
-                //  passed to the user on next recv call unless verbose mode is enabled
-                //  which makes to pass always these messages.
-                if (options.type == ZMQ_XPUB
-                    && (unique || (*data == 1 && verbose_subs)
-                        || (*data == 0 && verbose_unsubs && verbose_subs))) {
+                bool notify;
+                if (*data == 0) {
+                    mtrie_t::rm_result rm_result =
+                      subscriptions.rm (data + 1, size - 1, pipe_);
+                    //  TODO reconsider what to do if rm_result == mtrie_t::not_found
```

Comment:
This case will be hard to cover with the existing tests, given it can only happen if the sub socket thinks it has sent a subscription (since in the removal case xsub only sends a message if it can match it to its local table) but the xpub has not received it. Maybe fiddling around with HWM, but it could be flacky. Any ideas?

Reply:
This is probably also something that can better be tested with a kind of unit tests.

Can we defer this to after this PR?

Reply:
Sure we can - just worried about external behaviour changes, that's all. Do you want to do a rebase and squash?

Reply:
Can you push them as well? Can't see it

Reply:
Oops... should be up now